### PR TITLE
Add menu "Delete Recent Emojis

### PR DESCRIPTION
**Pull request by wizdom13**
Add a new feature in the conversation menu to delete recent emojis.

### CODE REVIEW

**Comment by agrajaghh** [comment]
I don't think moxie would like an option like that, see [CONTRIBUTING.md](https://github.com/WhisperSystems/Signal-Android/blob/master/CONTRIBUTING.md#development-ideology):
> The answer is not more options

Adding this option to the conversation menu feels way too intrusive. If this feature is desired, maybe a long-press popup menu on the recent smiley button, might be a better idea?

and I don't understand why you added the preference in `TextSecurePreferences` since you're just writing a `null` string and never read this preference?

**Comment by wizdom13** [comment]
Well, I think that form an organisation point of view that all the Preferences for the Textsecure should be in one place and hence TextSecurePreferences.

**Comment by agrajaghh** [comment]
but you are not saving anything there...

**Comment by wizdom13** [comment]
Regarding the place of the option, I did the same implementation of the WhatsApp Messenger.

**Comment by wizdom13** [comment]
If you guys prefer to move the DeleteRecentEmoji function, please let me know.

**Comment by agrajaghh** [comment]
btw, where are you deleting the recent emoijs? restarting the activity isn't deleting them, no?

**Comment by wizdom13** [comment]
This "setStringPreference(context, RECENT_EMOJI2_PREF, null);" delete the recent emoijs and restarting the activity is just for the reloading.

**Comment by agrajaghh** [comment]
omg, sorry, I just thought you introduced this since you added the constant, but its already in `RecentEmojiPageModel`. Forget what I said about that..

**Comment by wizdom13** [comment]
Never mind at all :)

**Comment by mcginty** [comment]
what's the reasoning for why someone would want to reset their recent emoji?

**Comment by wizdom13** [comment]
The user may want reset their recent emojis for various reasons mostly after using certain emojis and wanna start up clean. Beside it's a privacy issue, like if the user intend give or show Signal to some friend or family and he wanna keep his privacy. It was found in WhatsApp Messanger after so much requests from users.

**Comment by bungabunga** [comment]
@mcginty your wife goes to your Signal and finds that the recent emoji was heart, but she knows she didn't receive one...

**Comment by wizdom13** [comment]
@bungabunga @mcginty That's one example... 😂

**Comment by 2-4601** [comment]
But how's clearing the recent emojis going to stop @mcginty's wife from going through the conversations, opening one with @McLoo and seeing that it's full of (hairy) hearts?

**Comment by agrajaghh** [comment]
rofl... 
I guess he can delete the hairy heart messages like he already did with the heart itself! ;)

**Comment by moxie0** [comment]
thanks but I don't think we're going to do this

**Comment by wizdom13** [comment]
@moxie0

**Comment by wizdom13** [comment]
@moxie0 Can you explain please?

**Comment by moxie0** [comment]
i'd prefer to keep things simple, adding a new menu option isn't worth the value it provides in this case. in the future you can throw ideas out on the mailing list before writing the code to avoid wasting your time, and it'd be worth your while to go over the code style guidelines for future PRs as well.  thanks!

**Comment by wizdom13** [comment]
Yes you are right, it is a waste of time. I would rather put my efforts on something that will make its way out of the box. I see that this project with this mentality in projects management policies will not ever make this step. Good bye!

**Comment by cascheberg** [comment]
I also think that valid reasons been mentioned for why this feature should be available. But the specific implementation needs to be improved.
The solution @agrajaghh suggested is not permanently visible in the UI, but it can be found if users search for it.
An idea that came to my mind: add an option to the "Chats and media" settings that says "Do not save recently used emojis". That menu also contains the options to automatically delete old messages, which is somewhat similar.

**Comment by wizdom13** [comment]
@cascheberg listen to this. This application got very nice security and privacy ideas,  but it lacks so many options of the leading messagers and @moxie0 doesn't want to even talk about it. No one even can give any opinion or make any effort to make Signal better. I call this dictatorship in an open source project, thus I I'm thinking in forking the project and make a SignalPlus to make the same robust secure Signal with advanced and modern options. 
If any one do you it i will be a great supporter too.

**Comment by moxie0** [comment]
@wizdom13 This is an open source project, which means that you're free to do what you'd like with the source code subject to the terms of the license.  It *doesn't* mean that this is a democracy of the annoying, where the most annoying people on the planet have a contest of annoyingness to see who gets their way.

You can call it a dictatorship, but the decision making process in this project reflects the decision making process I'd like to see in the world: people's decision making influence is proportional to how they are effected by the outcome of the decision.

To break that down, I am the one who's ultimately stuck with this project -- I have to develop and maintain it for eternity.  You, on the other hand, showed up two days ago and quit a day later.  Given the relative long-term impact this code has on both of us, it seems pretty reasonable for me to be the one that makes the call about whether to include it or not.

To put it another way, you can rage quit this project whenever you want, but unfortunately there's no way for me to rage-quit you.

**Comment by wizdom13** [comment]
@moxie0 Well, I think accepting or refusing a new feature need more then 2 mins to decide and requires you to know the what are the opinion of the end users regarding this or at least other major active developers in the project.
@mcginty @agrajaghh @rhodey @veeti @unrulygnu @McLoo @thoughtbox @meskio @samlanning @haffenloher @tinloaf @phenx-de @cascheberg @turtlekiosk @backspace @AsamK @wsot @joeykrim @benalbrecht @kyleconroy @rymdhund @chr1shaefn3r @3xo @SamWhited

**Comment by samlanning** [comment]
@wizdom13 please don't mention people like you have just done. Most of the people in that list will not appreciate being mentioned in such a manner.

Mentions are for bringing people who are relevant into a conversation, not for bringing in the top 24 contributors of the project. (For the record, the majority of the people there will also agree with @moxie0 on this one).

Is this dictatorship in an open source project? Maybe... but then it's a benevolent dictatorship with well thought out goals and intentions. If the project hadn't been structured / managed as it is today, up until now... then it wouldn't be where it currently is, it wouldn't be nearly as popular as it is and user adoption wouldn't be as large as it is.

It's moxie's vision that has brought it here today.

Remember that democracy is not always strictly better than a dictatorship.

If you want to really get involved in this project and make a difference, I suggest that you take a step back and do some research and looking into what the goals of this project are, and why decisions are made the way they are. Then consider how your time and effort could be valuable / useful to the cause, and then submit a contribution that is part of the same vision.

Being an open source project does not mean that you are obliged to accept anyone's code, or that you even need to give a reason as to why you don't want to accept someone's code. Moxie gives better reasons than many project maintainers! There are those who just decide to swear at you because you capitalised a function wrong... you should take a look at the linux development mailing list and search for some of linus's responses... this is how open source works.

Sit back, take a look at some existing PRs, watch how things are done, then decide how you'd like to contribute.

**Comment by wizdom13** [comment]
No @samlanning I don't wanna be involved in this project. I already started my fork and it is now Signal Plus. Couple of days now and it already contains all the featured refused by @moxie0 to mention few:
   * Add Background color to chat conversations.
   * Add menu "Delete Recent Emojis"
   * Add BBM Emojis
   * Enable voice messages 
   * Few fixes and updates

Along with a "Chats Wallpaper" option that should be committed tomorrow. 
Please note that Signal Puls should be in Google Play within couple of weeks after changing the icon and few resources.

**Comment by wsot** [comment]
Commenting because I got pinged (although I've only committed two things for the project ever).

* Including BBM Emojis would be a bad idea because copyright.
* The argument for 'delete recent emojis' is unconvincing. If you're handing your phone to someone to show them Signal, I think you've given up any expectation of privacy. Maybe long-pressing on emojis should have that option, but it wouldn't make sense to put it anywhere else.
* I'm not interested in changing background colors, and even less about wallpapers, but I'm sure it would be nice for some people. Worth upping the minSdk version for though? I wouldn't think so.
* Other updates: well, I didn't see the pull requests.

Saying that someone is a more polite and helpful communicator than Linus is like saying that something tastes better than a rock, but Moxie (like others on the project) are well above that standard. It's Moxie's project first and foremost, and I am by and large comfortable with his decision making on his project to date.

**Comment by moxie0** [comment]
@wizdom13 You're free to use our code however you would like, but you're not free to use our name or our service.  If you feel that you can do better than us, that's fine, but please call your project something that won't confuse our users, and please run your own servers.  Thanks!
